flutter/flutter
# Part 2: Refactor Navigator's imperative API

This is a four part change:
1. https://github.com/flutter/flutter/pull/28747: Improves the Overlay API
2. (This change): refactors the Navigator's imperative API in preparation for part 3
3. (coming soon): Adds `Page`s to the Navigator
4. (coming soon): Adds a new `Router` widget

## Description

This change prepares the `Navigator` for the upcoming introduction of `Page`s, that can be used to declaratively set the content of the Navigator's history. This change moves the Navigator's imperative API (`push`, `pop`, `remove`, and friends) over to a model that will continue to work in that new world.

It also introduces a new callback `onGenerateInitialRoutes` on the Navigator, which can be used to change how the Navigator generates the initial Routes based on the `initialRoute` property.

The PR also contains clarifications to existing documentation and fixes some bugs:

* Routes generated from `initialRoute` will now correctly be marked as offstage when they are covered by an opaque route.
* Fixes https://github.com/flutter/flutter/issues/25080
* Fixes https://github.com/flutter/flutter/issues/16602
* Fixes https://github.com/flutter/flutter/issues/4172

## Related Issues

See above.

## Tests

I added the following tests:

* Ensure that `onGenerateInitialRoutes` is used correctly
* Ensure that routes pushed as part of `initialRoute` processing are marked offstage when covered by opaque routes
* Regression test for https://github.com/flutter/flutter/issues/25080.
* Regression tests for https://github.com/flutter/flutter/issues/16602.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.


## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [x] Yes, this is a breaking change (Please read [Handling breaking changes]):

* `Route.install` is no longer responsible for adding entries to the Navigator's overlay. The Navigator is responsible for this now. Therefore, `Route.install` will no longer be provided with a `overlayInsertionPoint`. Instead, it will get a boolean indicating if the route is an initial route, which the route may use to disable any entrance animations.
* For the same reasons outlined above, `TransitionRoute.createAnimationController` now takes a bool indicating if the Route is an initial route.
* `RouteSettings` no longer has an `initialRoute` property.
* `Navigator.pop` no longer returns a boolean, it's a void function now.
* A route is always disposed *after* all notifications concerning the route are sent to neighboring routes. Previously, sometimes the route would get disposed before these notifications have been sent.

A request for comment on those suggested breaking changes will be send out before this PR is considered for submission and after I've determined that part 3 of this endeavor doesn't require further breaking changes.


[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
